### ISSUE TITLE
EF Core include HasNoForeignKey in EntityTypeBuilder

### ISSUE THREAD

**Issue by emiliogg84** [dotnet/efcore#15046]
We need to create a semantic relation between two entities but not create a real foreign key in the database. These already work when chose "Database First" approach but I need to express the same behavior in "Code First" without the "Add-Migration" include a foreign key between the two tables in the database.

Example:

```C#
 public class CompanyAddress
    {
        public int ID { get; set; }
        [Required]
        public DateTimeOffset EffectiveDate { get; set; }
        [MaxLength(50)]
        public string AddressLine1 { get; set; }
        [MaxLength(50)]
        public string AddressLine2 { get; set; }
        public int CountryID { get; set; }        
        public string City { get; set; }
        public int? CountyID { get; set; }
        public int? ZipCodeID { get; set; }
        public bool IsDeleted { get; set; }
        public Guid CompanyID { get; set; }
        public Company Company { get; set; }
	**public int? StateID { get; set; }
        public State State { get; set; }**
    }

public class State
    {
        public int ID { get; set; }
        [MaxLength(3)]
        public string Code { get; set; }
        [MaxLength(100)]
        public string Name { get; set; }
        public int CountryID { get; set; }
        [JsonIgnore]
        public virtual Country Country { get; set; }
    }
```
I wan to add a migration that do not create a foreign key between "CompanyAddress" and "State" and the Include() function still work an load the state object for the "State" property of the "CompanyAddress"

In some point I want to do these: _context.Set<CompanyAddress>().Include(e => e.State); and EF must load the related entity.

A relation of One-two-One between "CompanyAddress" and "State" has not sense in the context of the "Data Model" (the database) neither in the context of the "Domain Model" is just a semantic relation that EF need to understand and load the related entity.

I thing that a new function calle HasNoForeignKey can solve the problem, these function will remove de foreign key instruction when the migration is create and... !!Every body happy!!

**Comment by ajcvickers** [comment]
@emililog84 Can you explain a little more why having a foreign key constraint in the database is bad for this scenario?

**Comment by emiliogg84** [comment]
EF Core when you add a migration adds a unique index with a foreign key in the above scenario due that the model match with a One-To-One relationship, but the State entity is well know as a classification entity ("a entity with simply data") that many entities in the model will reference to it.  

In the current project that I working we use ASP.NET Core (Web API) + EF Core + OData, the front-end team execute queries with OData that uses the $expand operator to display the State.Code value associate with the CompanyAddress.State property. As back-end team we not need a unique index and a  foreign key for this case, just that EF load the relate entity using the ID value backed in the CompanyAddress.StateId property to load the associate State object.

In other words... we need something that tell to the migration mechanism that not add the unique index and the foreign key. EF core still will load the associate value :-)

**Comment by ajcvickers** [comment]
@emiliogg84 Thanks for the additional info. The way to handle this is to edit the generated migrations to remove the code that creates FK constraint and index. (Scaffolded migrations are intended to be edited in this way for situations like this.)

We discussed doing something else here, but the scenario doesn't seem compelling and isn't something for which we have seen any other requests.

**Comment by emiliogg84** [comment]
Ok thant but... it will be nice to have a 'HasNoForeignKey' method to handle these situations!
